File: backend/src/services/categoryAnalyticsService.js
// src/services/categoryAnalyticsService.js
// Analítica de compras por categoria/produto — a base de dados/cálculo do
// Category Management (Economia de Escala) e do PO Robot (funcionalidade do
// add-on de automação, que reutiliza mediaMensalPorProduto em vez de
// duplicar o cálculo).
//
// Lê exclusivamente o que já existe (PurchaseOrderItem → Product →
// PurchaseOrder) — nenhuma tabela nova para isto. Considera só POs que
// representam compra reconhecida (mesmo universo do metricasService: PAGA em
// diante), para não contar ordens ainda por aprovar/pagar como volume.
const prisma = require('../config/database');

const RECONHECIDAS = ['PAGA', 'EM_EXECUCAO', 'ENTREGUE', 'RECEBIDA_CONFORME', 'RECEBIDA_COM_DIVERGENCIA', 'CONCLUIDA'];

function janela(dias = 365, agora = new Date()) {
  const de = new Date(agora);
  de.setDate(de.getDate() - dias);
  return { de, ate: agora };
}

const num = (d) => Number(d || 0);

/** Volume de compra reconhecida por categoria de produto, num período. */
async function volumePorCategoria({ companyId, de, ate } = {}) {
  const periodo = de && ate ? { de, ate } : janela(365);
  const itens = await prisma.purchaseOrderItem.findMany({
    where: {
      purchaseOrder: {
        buyerCompanyId: companyId,
        status: { in: RECONHECIDAS },
        createdAt: { gte: periodo.de, lte: periodo.ate },
      },
    },
    select: { lineTotal: true, product: { select: { category: true } } },
  });

  const porCategoria = new Map();
  let total = 0;
  for (const it of itens) {
    const categoria = it.product?.category || 'Sem categoria';
    const valor = num(it.lineTotal);
    porCategoria.set(categoria, (porCategoria.get(categoria) || 0) + valor);
    total += valor;
  }

  return {
    periodo: { de: periodo.de.toISOString(), ate: periodo.ate.toISOString() },
    total,
    categorias: [...porCategoria.entries()]
      .map(([categoria, valor]) => ({
        categoria,
        valor,
        percentual: total ? Math.round((valor / total) * 1000) / 10 : 0,
      }))
      .sort((a, b) => b.valor - a.valor),
  };
}

/** Quantidade comprada de um produto, agrupada por mês (YYYY-MM), ordenada. */
async function historicoMensalPorProduto({ companyId, productId, meses = 12 } = {}) {
  const { de, ate } = janela(meses * 30);
  const itens = await prisma.purchaseOrderItem.findMany({
    where: {
      productId,
      purchaseOrder: { buyerCompanyId: companyId, status: { in: RECONHECIDAS }, createdAt: { gte: de, lte: ate } },
    },
    select: { quantity: true, purchaseOrder: { select: { createdAt: true } } },
  });

  const porMes = new Map();
  for (const it of itens) {
    const d = it.purchaseOrder.createdAt;
    const chave = `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}`;
    porMes.set(chave, (porMes.get(chave) || 0) + it.quantity);
  }

  return [...porMes.entries()]
    .map(([mes, quantidade]) => ({ mes, quantidade }))
    .sort((a, b) => (a.mes < b.mes ? -1 : 1));
}

/**
 * Média mensal de compra de um produto, com base no histórico. Usada pelo
 * Category Management (recomendação) e pelo PO Robot (regra com
 * `mediaOrigem: 'IA'`) — a MESMA função, nunca duplicada entre as duas
 * funcionalidades.
 *
 * Divide-se pelos MESES DECORRIDOS desde a primeira compra do produto, não
 * só pelos meses em que houve compra: um produto comprado a cada 3 meses tem
 * uma média mensal de 1/3 da quantidade por compra, não a quantidade inteira
 * — senão o robot pediria a mesma quantidade todos os meses de um produto
 * que só se compra trimestralmente.
 */
async function mediaMensalPorProduto({ companyId, productId, meses = 12 } = {}) {
  const historico = await historicoMensalPorProduto({ companyId, productId, meses });
  if (!historico.length) {
    return { produtoId: productId, amostras: 0, mesesComCompra: 0, mediaMensal: 0 };
  }

  const quantidadeTotal = historico.reduce((s, m) => s + m.quantidade, 0);
  const [primeiroMes] = historico;
  const [ano0, mes0] = primeiroMes.mes.split('-').map(Number);
  const agora = new Date();
  const mesesDecorridos = Math.max(1, (agora.getFullYear() - ano0) * 12 + (agora.getMonth() + 1 - mes0) + 1);

  return {
    produtoId: productId,
    amostras: historico.length,
    mesesComCompra: historico.length,
    mesesDecorridos,
    quantidadeTotal,
    mediaMensal: Math.round((quantidadeTotal / mesesDecorridos) * 100) / 100,
  };
}

/**
 * Previsão simples (média móvel) da quantidade necessária no próximo mês,
 * a partir dos últimos `janelaMeses` com compra. Deliberadamente sem ML
 * pesado — é uma média, não um modelo, e diz-se isso ao cliente.
 */
async function previsaoNecessidade({ companyId, productId, janelaMeses = 3 } = {}) {
  const historico = await historicoMensalPorProduto({ companyId, productId, meses: 12 });
  if (!historico.length) return { produtoId: productId, previsaoProximoMes: 0, baseMeses: 0 };

  const ultimos = historico.slice(-janelaMeses);
  const media = ultimos.reduce((s, m) => s + m.quantidade, 0) / ultimos.length;
  return {
    produtoId: productId,
    previsaoProximoMes: Math.round(media * 100) / 100,
    baseMeses: ultimos.length,
  };
}

/**
 * Categorias com compra fragmentada (várias POs pequenas) que, consolidadas
 * num período, cruzariam o próximo threshold de desconto configurado.
 *
 * `thresholds` vem de quem chama (discountThresholdService.listar()) — esta
 * função não conhece os valores dos patamares, só compara contra o que lhe é
 * passado, para a configuração continuar a viver num único lugar.
 */
async function oportunidadesConsolidacao({ companyId, de, ate, thresholds } = {}) {
  const patamares = (thresholds || [])
    .filter((t) => t.ativo !== false)
    .map((t) => ({ minVolumeUsd: num(t.minVolumeUsd), discountPercent: num(t.discountPercent) }))
    .sort((a, b) => a.minVolumeUsd - b.minVolumeUsd);
  if (!patamares.length) return [];

  const periodo = de && ate ? { de, ate } : janela(365);
  const itens = await prisma.purchaseOrderItem.findMany({
    where: {
      purchaseOrder: {
        buyerCompanyId: companyId,
        status: { in: RECONHECIDAS },
        createdAt: { gte: periodo.de, lte: periodo.ate },
      },
    },
    select: { lineTotal: true, purchaseOrderId: true, product: { select: { category: true } } },
  });

  const porCategoria = new Map();
  for (const it of itens) {
    const categoria = it.product?.category || 'Sem categoria';
    if (!porCategoria.has(categoria)) porCategoria.set(categoria, { total: 0, pos: new Set() });
    const entrada = porCategoria.get(categoria);
    entrada.total += num(it.lineTotal);
    entrada.pos.add(it.purchaseOrderId);
  }

  const oportunidades = [];
  for (const [categoria, { total, pos }] of porCategoria.entries()) {
    const numeroPos = pos.size;
    // Exige fragmentação real: uma única PO grande já não é "consolidável".
    if (numeroPos < 3) continue;
    const proximo = patamares.find((t) => t.minVolumeUsd > total);
    if (!proximo) continue; // já está no maior patamar ativo
    const faltamUsd = proximo.minVolumeUsd - total;
    const mediaPorPo = total / numeroPos;
    // Só é uma "oportunidade" se o que falta é do tamanho de uma compra
    // típica desta categoria — não uma promessa distante e inatingível.
    if (faltamUsd > 0 && faltamUsd <= mediaPorPo) {
      oportunidades.push({
        categoria,
        volumeAtual: total,
        numeroPos,
        proximoThresholdUsd: proximo.minVolumeUsd,
        descontoPotencial: proximo.discountPercent,
        faltamUsd: Math.round(faltamUsd * 100) / 100,
      });
    }
  }
  return oportunidades.sort((a, b) => a.faltamUsd - b.faltamUsd);
}

module.exports = {
  volumePorCategoria,
  historicoMensalPorProduto,
  mediaMensalPorProduto,
  previsaoNecessidade,
  oportunidadesConsolidacao,
};

File: backend-java/src/main/java/ao/kixima/analytics/CategoryAnalyticsService.java
package ao.kixima.analytics;

import ao.kixima.po.PoStatus;
import ao.kixima.po.PurchaseOrderItemRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.Instant;
import java.time.YearMonth;
import java.time.ZoneOffset;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;

/**
 * Espelha só a parte de backend/src/services/categoryAnalyticsService.js que
 * o PO Robot reutiliza — {@code historicoMensalPorProduto} e
 * {@code mediaMensalPorProduto} (a MESMA função do Category Management,
 * nunca duplicada). Lê exclusivamente o que já existe (PurchaseOrderItem →
 * PurchaseOrder), só POs de compra reconhecida (PAGA em diante).
 *
 * NÃO PORTADO: volumePorCategoria/previsaoNecessidade/oportunidades de
 * consolidação — o dashboard de Category Management (ver
 * DiscountThresholdService, javadoc) fica para quando esse troço for portado.
 */
@Service
public class CategoryAnalyticsService {

    static final List<PoStatus> RECONHECIDAS = List.of(PoStatus.PAGA, PoStatus.EM_EXECUCAO, PoStatus.ENTREGUE,
            PoStatus.RECEBIDA_CONFORME, PoStatus.RECEBIDA_COM_DIVERGENCIA, PoStatus.CONCLUIDA);

    public record MesQuantidade(String mes, long quantidade) {
    }

    /** Espelha o objecto devolvido por mediaMensalPorProduto. */
    public record MediaMensal(String produtoId, int amostras, int mesesComCompra, Integer mesesDecorridos,
                              Long quantidadeTotal, BigDecimal mediaMensal) {
    }

    private final PurchaseOrderItemRepository purchaseOrderItemRepository;

    public CategoryAnalyticsService(PurchaseOrderItemRepository purchaseOrderItemRepository) {
        this.purchaseOrderItemRepository = purchaseOrderItemRepository;
    }

    /** Quantidade comprada de um produto, agrupada por mês (YYYY-MM), ordenada. */
    @Transactional(readOnly = true)
    public List<MesQuantidade> historicoMensalPorProduto(String companyId, String productId, int meses) {
        Instant ate = Instant.now();
        Instant de = ate.minus((long) meses * 30, ChronoUnit.DAYS);

        TreeMap<String, Long> porMes = new TreeMap<>();
        for (Object[] linha : purchaseOrderItemRepository.historicoDeCompra(companyId, productId, RECONHECIDAS, de, ate)) {
            YearMonth ym = YearMonth.from(((Instant) linha[0]).atZone(ZoneOffset.UTC));
            porMes.merge(ym.toString(), ((Number) linha[1]).longValue(), Long::sum);
        }

        List<MesQuantidade> historico = new ArrayList<>();
        porMes.forEach((mes, quantidade) -> historico.add(new MesQuantidade(mes, quantidade)));
        return historico;
    }

    /**
     * Média mensal de compra de um produto. Divide-se pelos MESES DECORRIDOS
     * desde a primeira compra, não só pelos meses em que houve compra — um
     * produto comprado a cada 3 meses tem média mensal de 1/3 da quantidade
     * por compra, senão o robot pediria a quantidade inteira todos os meses.
     */
    @Transactional(readOnly = true)
    public MediaMensal mediaMensalPorProduto(String companyId, String productId) {
        List<MesQuantidade> historico = historicoMensalPorProduto(companyId, productId, 12);
        if (historico.isEmpty()) {
            return new MediaMensal(productId, 0, 0, null, null, BigDecimal.ZERO);
        }

        long quantidadeTotal = historico.stream().mapToLong(MesQuantidade::quantidade).sum();
        YearMonth primeiro = YearMonth.parse(historico.get(0).mes());
        YearMonth agora = YearMonth.now(ZoneOffset.UTC);
        int mesesDecorridos = (int) Math.max(1, ChronoUnit.MONTHS.between(primeiro, agora) + 1);

        BigDecimal media = BigDecimal.valueOf(quantidadeTotal)
                .divide(BigDecimal.valueOf(mesesDecorridos), 2, RoundingMode.HALF_UP);
        return new MediaMensal(productId, historico.size(), historico.size(), mesesDecorridos, quantidadeTotal, media);
    }

    // --- Category Management (Lacunas D.7) ------------------------------------------

    public record Janela(Instant de, Instant ate) {
    }

    public static Janela janela(int dias) {
        Instant agora = Instant.now();
        return new Janela(agora.minus(dias, ChronoUnit.DAYS), agora);
    }

    private static double num(Object v) {
        return v == null ? 0 : ((Number) v).doubleValue();
    }

    /** Volume de compra reconhecida por categoria de produto, num período (365 dias por omissão). */
    @Transactional(readOnly = true)
    public Map<String, Object> volumePorCategoria(String companyId, Janela periodo) {
        Janela j = periodo != null ? periodo : janela(365);
        Map<String, Double> porCategoria = new LinkedHashMap<>();
        double total = 0;
        for (Object[] it : purchaseOrderItemRepository.linhasReconhecidasDaEmpresa(companyId, RECONHECIDAS, j.de(), j.ate())) {
            String categoria = it[0] == null ? "Sem categoria" : (String) it[0];
            double valor = num(it[1]);
            porCategoria.merge(categoria, valor, Double::sum);
            total += valor;
        }
        final double totalFinal = total;
        List<Map<String, Object>> categorias = new ArrayList<>();
        porCategoria.forEach((categoria, valor) -> {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("categoria", categoria);
            m.put("valor", valor);
            m.put("percentual", totalFinal > 0 ? Math.round((valor / totalFinal) * 1000) / 10.0 : 0d);
            categorias.add(m);
        });
        categorias.sort((a, b) -> Double.compare((double) b.get("valor"), (double) a.get("valor")));
        Map<String, Object> periodoOut = new LinkedHashMap<>();
        periodoOut.put("de", j.de());
        periodoOut.put("ate", j.ate());
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("periodo", periodoOut);
        out.put("total", total);
        out.put("categorias", categorias);
        return out;
    }

    /** Previsão simples (média móvel dos últimos `janelaMeses` com compra) — é uma média, não um modelo, e diz-se isso ao cliente. */
    @Transactional(readOnly = true)
    public Map<String, Object> previsaoNecessidade(String companyId, String productId, int janelaMeses) {
        List<MesQuantidade> historico = historicoMensalPorProduto(companyId, productId, 12);
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("produtoId", productId);
        if (historico.isEmpty()) {
            out.put("previsaoProximoMes", 0d);
            out.put("baseMeses", 0);
            return out;
        }
        List<MesQuantidade> ultimos = historico.subList(Math.max(0, historico.size() - janelaMeses), historico.size());
        double media = ultimos.stream().mapToLong(MesQuantidade::quantidade).sum() / (double) ultimos.size();
        out.put("previsaoProximoMes", Math.round(media * 100) / 100.0);
        out.put("baseMeses", ultimos.size());
        return out;
    }

    /** Um patamar de desconto como oportunidadesConsolidacao o recebe (vem de quem chama, nunca daqui). */
    public record Patamar(double minVolumeUsd, double discountPercent, boolean ativo) {
    }

    /** Categorias com compra fragmentada (≥ 3 POs) que, consolidadas, cruzariam o próximo patamar configurado. */
    @Transactional(readOnly = true)
    public List<Map<String, Object>> oportunidadesConsolidacao(String companyId, Janela periodo, List<Patamar> thresholds) {
        List<Patamar> patamares = new ArrayList<>();
        if (thresholds != null) for (Patamar t : thresholds) if (t.ativo()) patamares.add(t);
        patamares.sort(java.util.Comparator.comparingDouble(Patamar::minVolumeUsd));
        if (patamares.isEmpty()) return List.of();

        Janela j = periodo != null ? periodo : janela(365);
        Map<String, double[]> totais = new LinkedHashMap<>();
        Map<String, Set<String>> pos = new LinkedHashMap<>();
        for (Object[] it : purchaseOrderItemRepository.linhasReconhecidasDaEmpresa(companyId, RECONHECIDAS, j.de(), j.ate())) {
            String categoria = it[0] == null ? "Sem categoria" : (String) it[0];
            totais.computeIfAbsent(categoria, k -> new double[1])[0] += num(it[1]);
            pos.computeIfAbsent(categoria, k -> new HashSet<>()).add((String) it[2]);
        }
        List<Map<String, Object>> oportunidades = new ArrayList<>();
        for (Map.Entry<String, double[]> e : totais.entrySet()) {
            double total = e.getValue()[0];
            int numeroPos = pos.get(e.getKey()).size();
            if (numeroPos < 3) continue;
            Patamar proximo = patamares.stream().filter(t -> t.minVolumeUsd() > total).findFirst().orElse(null);
            if (proximo == null) continue; // já está no maior patamar ativo
            double faltamUsd = proximo.minVolumeUsd() - total;
            double mediaPorPo = total / numeroPos;
            if (faltamUsd > 0 && faltamUsd <= mediaPorPo) {
                Map<String, Object> m = new LinkedHashMap<>();
                m.put("categoria", e.getKey());
                m.put("volumeAtual", total);
                m.put("numeroPos", numeroPos);
                m.put("proximoThresholdUsd", proximo.minVolumeUsd());
                m.put("descontoPotencial", proximo.discountPercent());
                m.put("faltamUsd", Math.round(faltamUsd * 100) / 100.0);
                oportunidades.add(m);
            }
        }
        oportunidades.sort((a, b) -> Double.compare((double) a.get("faltamUsd"), (double) b.get("faltamUsd")));
        return oportunidades;
    }

    /** Espelha o objecto de mediaMensalPorProduto tal como a rota o devolve (mesmas chaves, incluindo as ausentes sem histórico). */
    public static Map<String, Object> mediaMensalOut(MediaMensal m) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("produtoId", m.produtoId());
        out.put("amostras", m.amostras());
        out.put("mesesComCompra", m.mesesComCompra());
        if (m.mesesDecorridos() != null) out.put("mesesDecorridos", m.mesesDecorridos());
        if (m.quantidadeTotal() != null) out.put("quantidadeTotal", m.quantidadeTotal());
        out.put("mediaMensal", m.mediaMensal());
        return out;
    }
}
